src/Corax/Utils/GrowableHashSet.cs

```diff
@@ -0,0 +1,74 @@
+﻿using System.Collections.Generic;
+
+namespace Corax.Utils;
+
+public sealed class GrowableHashSet<TItem>
+{
+    private List<HashSet<TItem>> _hashSetsBucket;
+    private HashSet<TItem> _newestHashSet;
+    private readonly int _maxSizePerCollection;
+    private readonly IEqualityComparer<TItem> _comparer;
+
+    public long Count
+    {
+        get
+        {
+            long result = _newestHashSet.Count;
+            if (_hashSetsBucket != null)
+            {
+                foreach (var hashBucket in _hashSetsBucket)
+                    result += hashBucket.Count;
+            }
+
+            return result;
+        }
+    }
+
+    public bool HasMultipleHashSets => _hashSetsBucket != null;
+
+    public GrowableHashSet(IEqualityComparer<TItem> comparer = null, int? maxSizePerCollection = null)
+    {
+        _comparer = comparer;
+        _hashSetsBucket = null;
+        _maxSizePerCollection = maxSizePerCollection ?? int.MaxValue;
+        CreateNewHashSet();
+    }
+
+    public bool Add(TItem item)
+    {
+        if (_newestHashSet!.Count >= _maxSizePerCollection)
```

Comment:
In order to reach here, we already have a 2B backing array, and we have a 100% used it.
So you have a lot of collisions.

Reply:
The default should be a max of 512M elements, I think.
Assuming load factor of 70% - that puts us under 1B elements. 
Now we'll need a new one here.